components/fxa-client/android/src/main/java/mozilla/appservices/fxa/Devices.kt

```diff
@@ -0,0 +1,83 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.fxa
+
+import mozilla.appservices.fxaclient.Device
+import mozilla.appservices.fxaclient.DeviceCapability
+import mozilla.appservices.fxaclient.DevicePushSubscription
+import mozilla.appservices.fxaclient.IncomingDeviceCommand
+import mozilla.appservices.sync15.DeviceType
+
+/**
+ * Manages the list of connected devices for a [FirefoxAccount]
+ */
+abstract class DeviceManager {
+    /**
+     * Set name of the current device.
+     * @param name New device name.
+     * @throwns FxaException.Authentication
+     */
+    abstract suspend fun setDeviceName(name: String)
+
+    /**
+     * Get the current device list. May be incomplete if state was never queried.
+     * @return [DeviceList] describes current and other known devices
+     */
+    abstract fun getKnownDevices(): DeviceList
+
+    /**
+     * Set a [DevicePushSubscription] for the current device.
+     * @param subscription A new [DevicePushSubscription].
+     * @throwns FxaException.Authentication
+     */
+    abstract suspend fun setDevicePushSubscription(subscription: DevicePushSubscription)
+
+    /**
+     * Send a command to a specified device.
+     * @param targetDeviceId A device ID of the recipient.
+     * @param title tab title
+     * @param url tab URL
+     * @throwns FxaException.Authentication
+     */
+    abstract suspend fun sendSingleTab(targetDeviceId: String, title: String, url: String)
+
+    /**
+     * Refreshes [DeviceList]. Registered [AccountEventsObserver] observers will be notified.
+     *
+     * @throwns FxaException.Authentication
+     */
+    abstract suspend fun refreshDevices()
+
+    /**
+     * Polls for any pending [DeviceCommandIncoming] commands.
+     * In case of new commands, registered [AccountEventsObserver] observers will be notified.
+     *
+     * @throwns FxaException.Authentication
+     */
+    abstract suspend fun pollForDeviceCommands()
+}
+
+/**
+ * Configuration for the current device.
+ *
+ * @property name An initial name to use for the device record which will be created during authentication.
+ * This can be changed later via [setDeviceName].
+ * @property type Type of a device - mobile, desktop - used for displaying identifying icons on other devices.
+ * This cannot be changed once device record is created.
+ * @property capabilities A set of device capabilities, such as SEND_TAB.
+ * @property secureStateAtRest A flag indicating whether or not to use encrypted storage for the persisted account
+ * state.
+ */
+data class DeviceConfig(
```

Comment:
Not quite sure of the scope of this PR - is it to describe a shape but not the impl, or is is something that would be checked in?

Specifically, I'm wondering why this `data class`, and the `enum` below, wouldn't be in UDL? And some of the functions below are duplicated - is that mainly because of the `suspend` attribute, and we'd ideally just remove those wrappers once uniffi could help, or are these wrappers a "feature"?

Reply:
It's more to describe the shape.  I don't want to write any actual code until Lougenia does her analysis with the Swift API and we can compare and contrast the two.

I consider the wrappers a necessary evil that's hopefully temporary.  My hope is that once we decide on an API, we implement almost all of the API in Rust and the Swift/Kotlin wrappers would just deal with the async parts.  And then those wrappers can fall off once UniFFI has a better async story.